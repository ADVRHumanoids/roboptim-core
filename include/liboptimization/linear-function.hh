// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the LinearFunction class.
 */

#ifndef OPTIMIZATION_LINEAR_FUNCTION_HH
# define OPTIMIZATION_LINEAR_FUNCTION_HH
# include <limits>
# include <liboptimization/quadratic-function.hh>

namespace optimization
{
  /// Define a linear function.
  class LinearFunction : public QuadraticFunction
  {
  public:
    LinearFunction (size_type n, size_type m = 1) throw ();

    virtual hessian_t hessian (const vector_t&, int) const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_LINEAR_FUNCTION_HH
