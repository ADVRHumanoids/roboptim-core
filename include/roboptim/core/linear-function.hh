// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_LINEAR_FUNCTION_HH
# define ROBOPTIM_CORE_LINEAR_FUNCTION_HH
# include <limits>
# include <roboptim/core/quadratic-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_meta_function
  /// @{

  /// \brief Define an abstract linear function.
  ///
  /// Inherit from this class when implementing linear functions.
  class LinearFunction : public QuadraticFunction
  {
  public:
    /// \brief Concrete class constructor should call this constructor.
    ///
    /// \param inputSize function arity
    /// \param outputSize result size
    LinearFunction (size_type inputSize, size_type outputSize = 1) throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();
  protected:
    void impl_hessian (hessian_t& hessian,
		       const argument_t& argument,
		       size_type functionId = 0) const throw ();
  };

  /// @}

} // end of namespace roboptim
#endif //! ROBOPTIM_CORE_LINEAR_FUNCTION_HH
