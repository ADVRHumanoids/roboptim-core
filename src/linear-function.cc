// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the LinearFunction class.
 */

#include "roboptim-core/linear-function.hh"
#include "roboptim-core/indent.hh"
#include "roboptim-core/util.hh"

namespace roboptim
{
  LinearFunction::LinearFunction (size_type n, size_type m) throw ()
    : QuadraticFunction (n, m)
  {
  }

  LinearFunction::hessian_t
  LinearFunction::hessian (const vector_t&, int) const throw ()
  {
    hessian_t h (n,n);
    h.clear ();
    return h;
  }

  std::ostream&
  LinearFunction::print (std::ostream& o) const throw ()
  {
    return o << "Linear function";
  }
} // end of namespace roboptim
