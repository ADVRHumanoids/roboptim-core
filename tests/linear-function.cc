// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/common.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/linear-function.hh>

using namespace roboptim;

struct Null : public LinearFunction
{
  Null () : LinearFunction (1, 1, "null function")
  {}

  void impl_compute (result_t& res, const argument_t&) const throw ()
  {
    res.clear ();
  }

  void impl_gradient (gradient_t& grad, const argument_t&,
		      size_type) const throw ()
  {
    grad.clear ();
  }
};

struct NoTitle : public LinearFunction
{
  NoTitle () : LinearFunction (1, 1)
  {}

  void impl_compute (result_t& res, const argument_t&) const throw ()
  {
    res.clear ();
  }

  void impl_gradient (gradient_t& grad, const argument_t&,
		      size_type) const throw ()
  {
    grad.clear ();
  }
};

int run_test ()
{
  Null null;
  NoTitle notitle;

  std::cout << null << std::endl
	    << notitle << std::endl;

  Null::vector_t x (1);
  x[0] = 42.;

  std::cout << null.gradient (x) << std::endl
	    << notitle.gradient (x) << std::endl;

  std::cout << null.hessian (x) << std::endl
	    << notitle.hessian (x) << std::endl;

  return 0;
}

GENERATE_TEST ()
