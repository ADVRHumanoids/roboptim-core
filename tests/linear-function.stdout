null function (linear function)
Linear function
[1](0)
[1](0)
[1](0)
[1](0)
